### ISSUE TITLE
traceback() for no applicable method for 'depth' applied to an object of class "NULL"

### ISSUE THREAD

**Issue by joethorley** [tidyverse/ggplot2#3538]
This error has been reported previously (in issue #2514) where @hadley wrote.

> Unfortunately there's not much I can do without a reprex. If you see it again, at least try capturing a `traceback()` and filing a new issue.

_Originally posted by @hadley in https://github.com/tidyverse/ggplot2/issues/2514#issuecomment-384983896_

It frequently but unpredictably occurs when printing ggplot objects by sourcing a script. 
As it's unpredictable I'm not able to provide a reprex however I am providing a traceback() as suggested.

```r
Error in UseMethod("depth") : 
  no applicable method for 'depth' applied to an object of class "NULL"
In addition: There were 13 warnings (use warnings() to see them)
> traceback()
16: depth(path)
15: upViewport(depth(vp), recording = FALSE)
14: popgrobvp.viewport(x$vp)
13: popgrobvp(x$vp)
12: postDraw.grob(x)
11: postDraw(x)
10: drawGTree(x)
9: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
8: grid.draw.gTree(gtable)
7: grid.draw(gtable)
6: print.ggplot(gp)
5: print(gp) at predict-abundance.R#138
4: eval(ei, envir)
3: eval(ei, envir)
2: withVisible(eval(ei, envir))
1: source("predict-abundance.R")
```

**Comment by joethorley** [comment]
I'm also willing to work with someone to try and narrow it down.

**Comment by yutannihilation** [comment]
The similar traceback on https://github.com/sdesabbata/BivariateTMap/issues/2 is of tmap, so it seems this error is not specific to ggplot2. I'm afraid there's nothing we can do...

**Comment by hadley** [comment]
I think this is something that @thomasp85 will eventually pick up as part of his general graphics/grid work.

**Comment by r2evans** [comment]
Since all of the error message I see are about methods and `NULL`, can we simply just add a `depth.NULL` S3 method?

```r
depth.NULL <- function(x, ...) 1
```

The current definitions of `grid::depth.*` ([github](https://github.com/wch/r-source/blob/5a156a0865362bb8381dcd69ac335f5174a4f60c/src/library/grid/R/viewport.R#L166-L194)) all appear to return numbers, so this seems reasonable. I don't know if `0` would be better than `1`, to be honest, but while I can see that 0 might make sense here (to mean: don't know, let's assume none), I don't know if `grid::` functions *expect* the depth to be 1+ or 0+.

**Comment by thomasp85** [comment]
@pmur002 do you have any insight as to what may cause this error? #2514 lists a range of observations about what may cause it... Could it potentially be related to garbage collection as it seems to disappear when the plot is assigned to a variable instead of just being drawn (note: plots shouldn't get garbage collected as they are kept in the the `.last_plot` object)

**Comment by thomasp85** [comment]
Side note... This error consistently comes up when checking patchwork on Ubuntu on R 3.2 - 3.5, but not on 3.6... was this fixed in the latest release?

https://github.com/thomasp85/patchwork/runs/319176273

**Comment by r2evans** [comment]
I can reproduce the error even when storing as a variable, so I believe that that is not the culprit (or at least the only one ... idk).

My steps to reproduce involve a multi-monitor extended desktop. Using just R (within emacs/ess):

1. Start a simple plot.

    ```r
    library(ggplot2)
    p <- ggplot(mtcars,aes(mpg,disp))+geom_point()
    p
    ```

2. "Snap" the window; I have reproduced the crash when it is maximized or full-height snapped onto either the left or right. (Because of this, I wonder with no verification whether the crash has something to do with Windows forcefully redefining desktop size that includes a pixel row/column on which the window resides.)

3. Disconnect the second monitor.

    ```r
    # Error in UseMethod("depth") : 
    #   no applicable method for 'depth' applied to an object of class "NULL"
    ```

During the crash, the R window goes blank, in the same fashion as when `ggplot2` clears the canvas during an plot draw or redraw.

Within RStudio, while I haven't tried it hundreds of times, I can repeat this most of the time:

- just the RSTudio plot window: does not crash;
- both the RStudio plot window and an external `windows` graphics device: does not crash, but takes longer than normal to redraw
- close the RStudio plot and keep the `windows` graphics device active: crash

    ![image](https://user-images.githubusercontent.com/5815808/69558733-9e463d80-0f5d-11ea-8e22-86bce4f3cd2b.png)

While this test uses a second monitor, it has also triggered consistently-enough without a second monitor when transitioning from a sleep/hibernate power state; if the premise of window-dimensions I offered above carries into this, then I believe windows might be altering the desktop geometry to allow the video devices to sleep ... though again, no verification. Testing to/from hibernation takes a lot longer, so I focused on reproducibility with a lot less time-requirement.

R-3.5.3, ggplot2-3.2.0

**Comment by r2evans** [comment]
I also just installed R-3.6.1 and ggplot2-3.2.1, and the same steps again continue to crash R, so it does not appear to be completely dependent on R < 3.6.

**Comment by pmur002** [comment]
Based on these descriptions and those in https://github.com/tidyverse/ggplot2/issues/2514 it sounds like it is related to a (mostly Windows or Mac ?) graphics device being resized or made visible.  That would trigger a replay of the graphics engine display list, so this is possibly to do with creating an unstable/corrupt 'grid' state by generating multiple overlapping replays of the graphics engine display list.  It could also be specific to how/when the Windows and/or Mac graphics device is triggering replays of the graphics engine display list.  If that wild speculation is anywhere near the mark, we are probably dealing with how these graphics devices handle events (like expose events).  There is evidence that both Windows and Quartz devices make use of a modal flag (xd->replaying and qd->redraw respectively) to avoid certain activities during a redraw.  Perhaps the problem relates to not making use of the modal flag when handling a particular event.  This gets complicated and requires both a reproducible example AND a debugging set up on Windows and/or Mac.  I do not have the latter at the moment, but may do so by next year, plus will have Simon Urbanek as a neighbour, so may be able to bug him about at least the Mac part :)

**Comment by thomasp85** [comment]
Good news is that I’ve managed to make a reproducible example. Bad news is that it is on Linux (in relation to your theory) and that I can only trigger it during a knitr build on R 3.5 and below... I’ll update once I know more

**Comment by r2evans** [comment]
@pmur002, your approach sounds solid, and I'm encouraged that somebody has been able to find a plausible theory for the underlying instability. I hope @thomasp85's `knitr`-reproducible example can provide some insight, as I know my "reproducibility" is lacking automation. Thanks! (This really is a frustrating bug that strikes without remorse and never at a convenient moment.)

**Comment by pmur002** [comment]
@thomasp85 , that is nice and reproducible for me, but ONLY for R < 3.6.  I believe I have tracked it down to a fix in grid (see doc/changes.txt) ...
```
Changes from grid_3.5.* to grid_3.6.0:
-------------------------------------
2.  Improve handling of graphics devices in grid.grabExpr().
    Should now work when multiple graphics devices already open
    (e.g., within 'knitr' document).
```
So unfortunately it seems that that is not the problem we are looking for.

**Comment by joethorley** [comment]
I'm getting the same error with R 4.0.0. It's so frequent this morning that I all 20+ attempts to rerun a script to produce some plots have failed...

A traceback is

```18: depth(path)
17: upViewport(depth(vp), recording = FALSE)
16: popgrobvp.viewport(x$vp)
15: popgrobvp(x$vp)
14: postDraw.grob(x)
13: postDraw(x)
12: drawGTree(x)
11: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
10: grid.draw.gTree(x$children[[i]], recording = FALSE)
9: NextMethod()
8: grid.draw.absoluteGrob(x$children[[i]], recording = FALSE)
7: grid.draw(x$children[[i]], recording = FALSE)
6: drawGTree(x)
5: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
4: grid.draw.gTree(gtable)
3: grid.draw(gtable)
2: print.ggplot(gp)
1: print(gp)```

**Comment by joethorley** [comment]
Here's another one 
```
26: depth(path)
25: upViewport(depth(vp), recording = FALSE)
24: popgrobvp.viewport(x$vp)
23: popgrobvp(x$vp)
22: postDraw.grob(x)
21: postDraw(x)
20: drawGTree(x)
19: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
18: grid.draw.gTree(x$children[[i]], recording = FALSE)
17: grid.draw(x$children[[i]], recording = FALSE)
16: drawGTree(x)
15: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
14: grid.draw.gTree(x$children[[i]], recording = FALSE)
13: grid.draw(x$children[[i]], recording = FALSE)
12: drawGTree(x)
11: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
10: grid.draw.gTree(x$children[[i]], recording = FALSE)
9: NextMethod()
8: grid.draw.absoluteGrob(x$children[[i]], recording = FALSE)
7: grid.draw(x$children[[i]], recording = FALSE)
6: drawGTree(x)
5: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
4: grid.draw.gTree(gtable)
3: grid.draw(gtable)
2: print.ggplot(gp)
1: print(gp)
```

**Comment by clauswilke** [comment]
@joethorley Tracebacks aren't particularly helpful. What we need is a reproducible example. If your script reliably fails, maybe you can turn it into a reprex? It's important to make it minimal, though. Remove everything that isn't relevant to triggering the problem.

**Comment by joethorley** [comment]
@clauswilke - the trouble is it's stochastic with the probability of an error changing from day to day. This morning the probability is really high - its somewhere in the region of a probability of 0.2 of failing for any given plot. Low for a single plot but when I try and run a script that produces over 25 plots the chance of it failing at some point is really high. I could screen share if that's helpful.

**Comment by r2evans** [comment]
@clauswilke , for the record, hadley's previous comment specifically suggested `traceback()` (in lieu of a specific reprex). While you did not say it the first time, I think having a traceback is still very *informative* even if it does not give you a reprex and a unit-test. (I've never been able to produce one because when this strikes me, it's always crashed my R process, completely unrecoverable. While I'm sure it's possible to attach `gdb` to R on windows, it's not something that is practical when the failure is non-deterministic, as joethorley said.)

Thanks for looking into it, I certainly am not trying to discourage involvement here. This is the first I've seen a traceback for this problem (that's plagued me for years), so at least there's some progress.

**Comment by joethorley** [comment]
FWIW it never crashes my R process - I'm running in RStudio with OSX. For some reason my machine is hot this morning - I have a script that produces enough plots that one of them is almost guaranteed to fail. If someone with more expertise is willing to screen share with me on Zoom we may be able to make more progress in understanding what is happening?

**Comment by r2evans** [comment]
@clauswilke, is there a way to get "mouse actions" in a reprex? I can get a crash relatively reliably by plotting a simple one (literally `ggplot(mtcars,aes(mpg,disp))+geom_point()`) and mouse-scrolling (like a *mad*-man) in the graphics window. With repeated errors such as:

```r
Error in grid.Call.graphics(C_setviewport, vp, TRUE) : 
  REAL() can only be applied to a 'numeric', not a 'NULL'
Error during wrapup: no applicable method for 'depth' applied to an object of class "NULL"
Error in grid.Call.graphics(C_setviewport, vp, TRUE) : 
  VECTOR_ELT() can only be applied to a 'list', not a 'character'
Error during wrapup: REAL() can only be applied to a 'numeric', not a 'character'
Error in grid.Call.graphics(C_upviewport, as.integer(n)) : 
  cannot pop the top-level viewport ('grid' and 'graphics' output mixed?)
Error during wrapup: no applicable method for 'depth' applied to an object of class "NULL"
Error in grid.Call.graphics(C_upviewport, as.integer(n)) : 
  cannot pop the top-level viewport ('grid' and 'graphics' output mixed?)
Error during wrapup: no applicable method for 'depth' applied to an object of class "NULL"

Process R:2 finished at Fri May  1 10:33:29 2020
```

It's not a programmatic reprex, but I've done it three times in a row (mousewheel-scrolling like *crazy*). This is the best way I have so far to reproduce many of the errors that usually result in a lost window and/or crashed R.

**Comment by r2evans** [comment]
Frankly, I'm not convinced that this is wholly within `ggplot2`. The error message

```r
Error: REAL() can only be applied to a 'numeric', not a 'NULL'
```

to me strongly suggests a C component, which `ggplot2` itself doesn't use. Its imports include `grid` and `grDevices`, which both use C files and the `REAL(...)` call. Unfortunately, this error does *not* provide any traceback information, so there's no easy way to know which call is passing `NULL`. Since all of my graphics-related crashes have been while using `ggplot2`, my lean is towards `grid` (between the two), though there is a lot of `REAL(...)` in that package ([github search](https://github.com/search?l=C&q=%22REAL%22+repo%3Awch%2Fr-source+path%3Asrc%2Flibrary%2Fgrid&type=Code)).

**Comment by pmur002** [comment]
I think that it is extremely likely to be in 'grid' - unfortunately, still need a reliable reprex (on Linux) that I can debug.

**Comment by joethorley** [comment]
I've since done a new install on a brand new Mac. The error is

```r
Error in grid.Call.graphics(C_upviewport, as.integer(n)) : 
  cannot pop the top-level viewport ('grid' and 'graphics' output mixed?)
In addition: There were 12 warnings (use warnings() to see them)
Backtrace:
     █
  1. ├─[ base::source(...) ] with 3 more calls
  5. ├─[ global::plot_year_effect(...) ] with 6 more calls predict-growth.R:35:2
 12. ├─[ grid:::drawGTree(...) ] with 5 more calls
 18. ├─[ grid:::drawGTree(...) ] with 13 more calls
 32. ├─[ (function (x) ... ]
 33. └─grid:::postDraw.grob(...)
 34.   ├─[ grid:::popgrobvp(...) ]
 35.   └─grid:::popgrobvp.viewport(x$vp)
 36.     └─grid::upViewport(depth(vp), recording = FALSE)
 37.       └─grid:::grid.Call.graphics(C_upviewport, as.integer(n))
> 
```

**Comment by joethorley** [comment]
FWIW I now use the following code to print ggplot2 objects.

```
grid_errors <- function() {
  c("cannot pop the top-level viewport", 
    "no applicable method for 'depth'")
}

sbf_print <- function(x, ntry = 3L, silent = FALSE) {
  i <- 1
  while(i < ntry) {
    try <- try(print(x), silent = TRUE)
    if(!vld_is(try, "try-error"))
      return(invisible(try))
    if(!any(vapply(grid_errors(), grepl, TRUE, x = try))) {
      if(!silent) cat(x)
      return(print(x))
    }
    i <- i + 1
  }
  print(x)
}```

**Comment by pmur002** [comment]
I have gained access to a macOS machine - did anyone have a code-based reproducible example that I could try out ?

**Comment by joethorley** [comment]
``` r
library(ggplot2)

for(i in 1:40) {
  print(i)
  gp <- ggplot(data = mtcars, aes(x = wt, y = mpg)) +
    geom_point()
  quartz()
  print(gp)
}
#> [1] 1
#> [1] 2
#> [1] 3
#> Error in UseMethod("depth"): no applicable method for 'depth' applied to an object of class "NULL"
```

<sup>Created on 2020-05-11 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by pmur002** [comment]
Thanks.  That does it for me.  Now, let's see if I can see why ...

**Comment by pmur002** [comment]
I believe I have tracked the source of this problem and, as suspected, it is (in the quartz() case at least) the device triggering a redraw in the middle of drawing (presumably in response to a window event of some sort), which reinitialises the 'grid' system state, including the 'grid' viewport stack, and everything crashes to a halt.  So some progress, but I now need to enlist some help from the quartz() device author(s) to try to find a fix.

**Comment by r2evans** [comment]
*Thank you*, @pmur002, for doing what I don't have the proficiency to do myself. Thank you.

**Comment by pmur002** [comment]
Let's wait and see if I have the proficiency :)  More likely, Simon Urbanek is going to have the proficiency!

**Comment by TobieSurette** [comment]
I've just updated to R 4.0.2 and RStudio 1.3.959 on my MacOS 10.14.1. I'm having some very perplexing issues with all my S3 object classes and it's reporting a very similar error as described above. This is probably a Mac issue, so I'm not sure that this bug arises in Windows. 

I've reduced it to its simplest form:

Say we define two methods in an S3 class:
```
foo <- function(x, ...) UseMethod("foo")
foo.numeric <- function(x, ...) return(2*x)
```
and then run `foo(3)`. Then it correctly returns a `6`. However, when I define the same functions in their own package, then load the package (i.e. `library(foo)`) in a new R session and try `foo(3)`, it returns this message:

```
Error in UseMethod("foo") : 
  no applicable method for 'foo' applied to an object of class "c('double', 'numeric')"
```
Even more odd, the files seem to be defined in the environment, so typing `foo` and `foo.default` correctly return:
```
function(x, ...) UseMethod("foo")
<bytecode: 0x7feae1163778>
<environment: namespace:foo>
```
and 
```
function(x, ...) return(2*x)
<bytecode: 0x7feae198a788>
<environment: namespace:foo>
```
Any help would be appreciated, 
Tobie

**Comment by yutannihilation** [comment]
@TobieSurette It sounds a very different issue about S3 method lookups. You probably need to register your method to make it visible to other environments than the global.

**Comment by TobieSurette** [comment]
Also, the call to the generic returns:

Error in UseMethod("foo") : 
  no applicable method for 'foo' applied to an object of class "NULL"

**Comment by TobieSurette** [comment]
Thank you, I'll check it out

**Comment by r2evans** [comment]
@pmur002, update: I'm now on R-4.0.2 and `ggplot2-3.3.0` (still win10) and while it is much *easier* now to trigger the response (just mouse over a `ggplot` graphic window and scroll the mouse wheel), it seems to crash a little less often. 

<sub>*Not sure if it helps any, I'm trying to add whatever perspective I can to the problem, hoping somebody will have an "aha!" moment.*</sub>

**Comment by sanjmeh** [comment]
Let me also throw my hat into this and try help diagnose the warning:
```
Error in UseMethod("depth") : 
  no applicable method for 'depth' applied to an object of class "NULL"
```

The problem started very recently, after I wrote a ggplot function and added to my R script source. After converting the warning to an error (Debug -> On Error -> Break In Code, the traceback looks like this.
```
> traceback()
32: depth(path)
31: upViewport(depth(vp), recording = FALSE)
30: popgrobvp.viewport(x$vp)
29: popgrobvp(x$vp)
28: postDraw.grob(structure(list(widths = structure(list(list(0, 
        NULL, 8L), list(1, structure(list(list(1, structure(list(
        label = c("101", "203", "204", "302", "303", "401", "501", 
        "502", "503", "504", "602", "604", "702", "704", "804", "903", 
        "904", "1002", "1102", "1204", "1402"), x = structure(c(1, 
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
        1), unit = 0L, class = c("simpleUnit", "unit", "unit_v2")), 
        y = structure(c(0.0283018867924528, 0.0754716981132075, 0.122641509433962, 
        0.169811320754717, 0.216981132075472, 0.264150943396226, 
        0.311320754716981, 0.358490566037736, 0.40566037735849, 0.452830188679245, 
        0.5, 0.547169811320755, 0.594339622641509, 0.641509433962264, 
        0.688679245283019, 0.735849056603773, 0.783018867924528, 
        0.830188679245283, 0.877358490566038, 0.924528301886792, 
        0.971698113207547), unit = 4L, class = c("simpleUnit", "unit", 
        "unit_v2")), just = "centre", hjust = 1, vjust = 0.5, rot = 0, 
        check.overlap = FALSE, name = "GRID.text.6717", gp = structure(list(
            fontsize = 8.8, col = "grey30", fontfamily = "", lineheight = 0.9, 
            font = c(plain = 1L)), class = "gpar"), vp = NULL), class = c("text", 
    "grob", "gDesc")), 21L), list(0, NULL, 2L)), class = c("unit", 
    "unit_v2")), 201L), list(2.2, NULL, 8L)), class = c("unit", "unit_v2"
    )), heights = structure(1, unit = 5L, class = c("simpleUnit", 
    "unit", "unit_v2")), name = "GRID.titleGrob.6718", gp = NULL, 
        vp = structure(list(parent = structure(list(x = structure(0.5, unit = 0L, class = c("simpleUnit", 
        "unit", "unit_v2")), y = structure(0.5, unit = 0L, class = c("simpleUnit", 
        "unit", "unit_v2")), width = structure(1, unit = 0L, class = c("simpleUnit", 
        "unit", "unit_v2")), height = structure(1, unit = 0L, class = c("simpleUnit", 
        "unit", "unit_v2")), justification = "centre", gp = structure(list(
            fontsize = 8.8, col = "grey30", fontfamily = "", lineheight = 0.9, 
            font = c(plain = 1L)), class = "gpar"), clip = FALSE, 
            xscale = c(0, 1), yscale = c(0, 1), angle = 0, layout = structure(list(
                nrow = 1L, ncol = 3L, widths = structure(list(list(
                    0, NULL, 8L), list(1, structure(list(list(1, 
                    structure(list(label = c("101", "203", "204", 
                    "302", "303", "401", "501", "502", "503", "504", 
                    "602", "604", "702", "704", "804", "903", "904", 
                    "1002", "1102", "1204", "1402"), x = structure(c(1, 
                    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
                    1, 1, 1, 1, 1), unit = 0L, class = c("simpleUnit", 
                    "unit", "unit_v2")), y = structure(c(0.0283018867924528, 
                    0.0754716981132075, 0.122641509433962, 0.169811320754717, 
                    0.216981132075472, 0.264150943396226, 0.311320754716981, 
                    0.358490566037736, 0.40566037735849, 0.452830188679245, 
                    0.5, 0.547169811320755, 0.594339622641509, 0.641509433962264, 
                    0.688679245283019, 0.735849056603773, 0.783018867924528, 
                    0.830188679245283, 0.877358490566038, 0.924528301886792, 
                    0.971698113207547), unit = 4L, class = c("simpleUnit", 
                    "unit", "unit_v2")), just = "centre", hjust = 1, 
                      vjust = 0.5, rot = 0, check.overlap = FALSE, 
                      name = "GRID.text.6717", gp = structure(list(
                        fontsize = 8.8, col = "grey30", fontfamily = "", 
                        lineheight = 0.9, font = c(plain = 1L)), class = "gpar"), 
                      vp = NULL), class = c("text", "grob", "gDesc"
                    )), 21L), list(0, NULL, 2L)), class = c("unit", 
                "unit_v2")), 201L), list(2.2, NULL, 8L)), class = c("unit", 
                "unit_v2")), heights = structure(1, unit = 5L, class = c("simpleUnit", 
                "unit", "unit_v2")), respect = FALSE, valid.respect = 0L, 
                respect.mat = structure(c(0L, 0L, 0L), .Dim = c(1L, 
                3L)), just = "centre", valid.just = c(0.5, 0.5)), class = "layout"), 
            layout.pos.row = NULL, layout.pos.col = NULL, valid.just = c(0.5, 
            0.5), valid.pos.row = NULL, valid.pos.col = NULL, name = "GRID.VP.2372"), class = "viewport"), 
            children = structure(list(structure(list(x = structure(0.5, unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), y = structure(0.5, unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), width = structure(1, unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), height = structure(1, unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), justification = "centre", gp = structure(list(), class = "gpar"), 
                clip = FALSE, xscale = c(0, 1), yscale = c(0, 1), 
                angle = 0, layout = NULL, layout.pos.row = NULL, 
                layout.pos.col = c(2L, 2L), valid.just = c(0.5, 0.5
                ), valid.pos.row = NULL, valid.pos.col = c(2L, 2L
                ), name = "GRID.VP.2373"), class = "viewport")), class = c("vpList", 
            "viewport"))), class = c("vpTree", "viewport")), children = structure(list(
            GRID.text.6717 = structure(list(label = c("101", "203", 
            "204", "302", "303", "401", "501", "502", "503", "504", 
            "602", "604", "702", "704", "804", "903", "904", "1002", 
            "1102", "1204", "1402"), x = structure(c(1, 1, 1, 1, 
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1), unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), y = structure(c(0.0283018867924528, 
            0.0754716981132075, 0.122641509433962, 0.169811320754717, 
            0.216981132075472, 0.264150943396226, 0.311320754716981, 
            0.358490566037736, 0.40566037735849, 0.452830188679245, 
            0.5, 0.547169811320755, 0.594339622641509, 0.641509433962264, 
            0.688679245283019, 0.735849056603773, 0.783018867924528, 
            0.830188679245283, 0.877358490566038, 0.924528301886792, 
            0.971698113207547), unit = 4L, class = c("simpleUnit", 
            "unit", "unit_v2")), just = "centre", hjust = 1, vjust = 0.5, 
                rot = 0, check.overlap = FALSE, name = "GRID.text.6717", 
                gp = structure(list(fontsize = 8.8, col = "grey30", 
                    fontfamily = "", lineheight = 0.9, font = c(plain = 1L)), class = "gpar"), 
                vp = NULL), class = c("text", "grob", "gDesc"))), class = "gList"), 
        childrenOrder = "GRID.text.6717"), class = c("titleGrob", 
    "gTree", "grob", "gDesc")))
27: (function (x) 
    {
        UseMethod("postDraw")
    })(structure(list(widths = structure(list(list(0, NULL, 8L), 
        list(1, structure(list(list(1, structure(list(label = c("101", 
        "203", "204", "302", "303", "401", "501", "502", "503", "504", 
        "602", "604", "702", "704", "804", "903", "904", "1002", 
        "1102", "1204", "1402"), x = structure(c(1, 1, 1, 1, 1, 1, 
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1), unit = 0L, class = c("simpleUnit", 
        "unit", "unit_v2")), y = structure(c(0.0283018867924528, 
        0.0754716981132075, 0.122641509433962, 0.169811320754717, 
        0.216981132075472, 0.264150943396226, 0.311320754716981, 
        0.358490566037736, 0.40566037735849, 0.452830188679245, 0.5, 
        0.547169811320755, 0.594339622641509, 0.641509433962264, 
        0.688679245283019, 0.735849056603773, 0.783018867924528, 
        0.830188679245283, 0.877358490566038, 0.924528301886792, 
        0.971698113207547), unit = 4L, class = c("simpleUnit", "unit", 
        "unit_v2")), just = "centre", hjust = 1, vjust = 0.5, rot = 0, 
            check.overlap = FALSE, name = "GRID.text.6717", gp = structure(list(
                fontsize = 8.8, col = "grey30", fontfamily = "", 
                lineheight = 0.9, font = c(plain = 1L)), class = "gpar"), 
            vp = NULL), class = c("text", "grob", "gDesc")), 21L), 
            list(0, NULL, 2L)), class = c("unit", "unit_v2")), 201L), 
        list(2.2, NULL, 8L)), class = c("unit", "unit_v2")), heights = structure(1, unit = 5L, class = c("simpleUnit", 
    "unit", "unit_v2")), name = "GRID.titleGrob.6718", gp = NULL, 
        vp = structure(list(parent = structure(list(x = structure(0.5, unit = 0L, class = c("simpleUnit", 
        "unit", "unit_v2")), y = structure(0.5, unit = 0L, class = c("simpleUnit", 
        "unit", "unit_v2")), width = structure(1, unit = 0L, class = c("simpleUnit", 
        "unit", "unit_v2")), height = structure(1, unit = 0L, class = c("simpleUnit", 
        "unit", "unit_v2")), justification = "centre", gp = structure(list(
            fontsize = 8.8, col = "grey30", fontfamily = "", lineheight = 0.9, 
            font = c(plain = 1L)), class = "gpar"), clip = FALSE, 
            xscale = c(0, 1), yscale = c(0, 1), angle = 0, layout = structure(list(
                nrow = 1L, ncol = 3L, widths = structure(list(list(
                    0, NULL, 8L), list(1, structure(list(list(1, 
                    structure(list(label = c("101", "203", "204", 
                    "302", "303", "401", "501", "502", "503", "504", 
                    "602", "604", "702", "704", "804", "903", "904", 
                    "1002", "1102", "1204", "1402"), x = structure(c(1, 
                    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 
                    1, 1, 1, 1, 1), unit = 0L, class = c("simpleUnit", 
                    "unit", "unit_v2")), y = structure(c(0.0283018867924528, 
                    0.0754716981132075, 0.122641509433962, 0.169811320754717, 
                    0.216981132075472, 0.264150943396226, 0.311320754716981, 
                    0.358490566037736, 0.40566037735849, 0.452830188679245, 
                    0.5, 0.547169811320755, 0.594339622641509, 0.641509433962264, 
                    0.688679245283019, 0.735849056603773, 0.783018867924528, 
                    0.830188679245283, 0.877358490566038, 0.924528301886792, 
                    0.971698113207547), unit = 4L, class = c("simpleUnit", 
                    "unit", "unit_v2")), just = "centre", hjust = 1, 
                      vjust = 0.5, rot = 0, check.overlap = FALSE, 
                      name = "GRID.text.6717", gp = structure(list(
                        fontsize = 8.8, col = "grey30", fontfamily = "", 
                        lineheight = 0.9, font = c(plain = 1L)), class = "gpar"), 
                      vp = NULL), class = c("text", "grob", "gDesc"
                    )), 21L), list(0, NULL, 2L)), class = c("unit", 
                "unit_v2")), 201L), list(2.2, NULL, 8L)), class = c("unit", 
                "unit_v2")), heights = structure(1, unit = 5L, class = c("simpleUnit", 
                "unit", "unit_v2")), respect = FALSE, valid.respect = 0L, 
                respect.mat = structure(c(0L, 0L, 0L), .Dim = c(1L, 
                3L)), just = "centre", valid.just = c(0.5, 0.5)), class = "layout"), 
            layout.pos.row = NULL, layout.pos.col = NULL, valid.just = c(0.5, 
            0.5), valid.pos.row = NULL, valid.pos.col = NULL, name = "GRID.VP.2372"), class = "viewport"), 
            children = structure(list(structure(list(x = structure(0.5, unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), y = structure(0.5, unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), width = structure(1, unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), height = structure(1, unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), justification = "centre", gp = structure(list(), class = "gpar"), 
                clip = FALSE, xscale = c(0, 1), yscale = c(0, 1), 
                angle = 0, layout = NULL, layout.pos.row = NULL, 
                layout.pos.col = c(2L, 2L), valid.just = c(0.5, 0.5
                ), valid.pos.row = NULL, valid.pos.col = c(2L, 2L
                ), name = "GRID.VP.2373"), class = "viewport")), class = c("vpList", 
            "viewport"))), class = c("vpTree", "viewport")), children = structure(list(
            GRID.text.6717 = structure(list(label = c("101", "203", 
            "204", "302", "303", "401", "501", "502", "503", "504", 
            "602", "604", "702", "704", "804", "903", "904", "1002", 
            "1102", "1204", "1402"), x = structure(c(1, 1, 1, 1, 
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1), unit = 0L, class = c("simpleUnit", 
            "unit", "unit_v2")), y = structure(c(0.0283018867924528, 
            0.0754716981132075, 0.122641509433962, 0.169811320754717, 
            0.216981132075472, 0.264150943396226, 0.311320754716981, 
            0.358490566037736, 0.40566037735849, 0.452830188679245, 
            0.5, 0.547169811320755, 0.594339622641509, 0.641509433962264, 
            0.688679245283019, 0.735849056603773, 0.783018867924528, 
            0.830188679245283, 0.877358490566038, 0.924528301886792, 
            0.971698113207547), unit = 4L, class = c("simpleUnit", 
            "unit", "unit_v2")), just = "centre", hjust = 1, vjust = 0.5, 
                rot = 0, check.overlap = FALSE, name = "GRID.text.6717", 
                gp = structure(list(fontsize = 8.8, col = "grey30", 
                    fontfamily = "", lineheight = 0.9, font = c(plain = 1L)), class = "gpar"), 
                vp = NULL), class = c("text", "grob", "gDesc"))), class = "gList"), 
        childrenOrder = "GRID.text.6717"), class = c("titleGrob", 
    "gTree", "grob", "gDesc")))
26: grid.Call.graphics(C_setviewport, vp, TRUE)
25: push.vp.viewport(X[[i]], ...)
24: FUN(X[[i]], ...)
23: lapply(vps, push.vp, recording)
22: pushViewport(vp, recording = FALSE)
21: pushgrobvp.viewport(x$vp)
20: pushgrobvp(x$vp)
19: pushvpgp(x)
18: preDraw.gTree(x)
17: preDraw(x)
16: drawGTree(x)
15: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
14: grid.draw.gTree(x$children[[i]], recording = FALSE)
13: grid.draw(x$children[[i]], recording = FALSE)
12: drawGTree(x)
11: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
10: grid.draw.gTree(x$children[[i]], recording = FALSE)
9: NextMethod()
8: grid.draw.absoluteGrob(x$children[[i]], recording = FALSE)
7: grid.draw(x$children[[i]], recording = FALSE)
6: drawGTree(x)
5: recordGraphics(drawGTree(x), list(x = x), getNamespace("grid"))
4: grid.draw.gTree(gtable)
3: grid.draw(gtable)
2: print.ggplot(x)
1: (function (x, ...) 
   UseMethod("print"))(x)
```

My session info is:

```
 setting  value                       
 version  R version 4.0.0 (2020-04-24)
 os       macOS Catalina 10.15.5      
 system   x86_64, darwin17.0          
 ui       RStudio                     
 language (EN)                        
 collate  en_GB.UTF-8                 
 ctype    en_GB.UTF-8                 
 tz       Asia/Kolkata                
 date     2020-07-15   
```

And package info (reduced it to show relevant ones):

```
 package         * version  date       lib source        
 anytime         * 0.3.7    2020-01-20 [1] CRAN (R 4.0.0)
 askpass           1.1      2019-01-13 [1] CRAN (R 4.0.0)
 assertthat      * 0.2.1    2019-03-21 [1] CRAN (R 4.0.0)
 backports         1.1.8    2020-06-17 [1] CRAN (R 4.0.0)
 bit             * 1.1-15.2 2020-02-10 [1] CRAN (R 4.0.0)
 bit64           * 0.9-7    2017-05-08 [1] CRAN (R 4.0.0)
 boot              1.3-24   2019-12-20 [1] CRAN (R 4.0.0)
 cellranger      * 1.1.0    2016-07-27 [1] CRAN (R 4.0.0)
 checkmate         2.0.0    2020-02-06 [1] CRAN (R 4.0.0)
 cli               2.0.2    2020-02-28 [1] CRAN (R 4.0.0)
 colorspace        1.4-1    2019-03-18 [1] CRAN (R 4.0.0)
 crayon            1.3.4    2017-09-16 [1] CRAN (R 4.0.0)
 curl              4.3      2019-12-02 [1] CRAN (R 4.0.0)
 data.table      * 1.12.8   2019-12-09 [1] CRAN (R 4.0.0)
 desc              1.2.0    2018-05-01 [1] CRAN (R 4.0.0)
 DescTools       * 0.99.35  2020-04-30 [1] CRAN (R 4.0.0)
 digest            0.6.25   2020-02-23 [1] CRAN (R 4.0.0)
 dplyr           * 1.0.0    2020-05-29 [1] CRAN (R 4.0.0)
 ellipsis          0.3.1    2020-05-15 [1] CRAN (R 4.0.0)
 evaluate          0.14     2019-05-28 [1] CRAN (R 4.0.0)
 expm              0.999-4  2019-03-21 [1] CRAN (R 4.0.0)
 fansi             0.4.1    2020-01-08 [1] CRAN (R 4.0.0)
 farver            2.0.3    2020-01-16 [1] CRAN (R 4.0.0)
 formattable     * 0.2.0.1  2016-08-05 [1] CRAN (R 4.0.0)
 fs                1.4.1    2020-04-04 [1] CRAN (R 4.0.0)
 gargle            0.5.0    2020-05-06 [1] CRAN (R 4.0.0)
 generics          0.0.2    2018-11-29 [1] CRAN (R 4.0.0)
 ggplot2         * 3.3.2    2020-06-19 [1] CRAN (R 4.0.0)
 glue              1.4.1    2020-05-13 [1] CRAN (R 4.0.0)
 googledrive     * 1.0.1    2020-05-05 [1] CRAN (R 4.0.0)
 googlesheets    * 0.3.0    2018-06-29 [1] CRAN (R 4.0.0)
 googlesheets4   * 0.2.0    2020-05-08 [1] CRAN (R 4.0.0)
 gtable            0.3.0    2019-03-25 [1] CRAN (R 4.0.0)
 hms               0.5.3    2020-01-08 [1] CRAN (R 4.0.0)
 htmlTable       * 1.13.3   2019-12-04 [1] CRAN (R 4.0.0)
 htmltools       * 0.4.0    2019-10-04 [1] CRAN (R 4.0.0)
 htmlwidgets       1.5.1    2019-10-08 [1] CRAN (R 4.0.0)
 httr              1.4.1    2019-08-05 [1] CRAN (R 4.0.0)
 janitor           2.0.1    2020-04-12 [1] CRAN (R 4.0.0)
 jsonlite          1.7.0    2020-06-25 [1] CRAN (R 4.0.0)
 kableExtra      * 1.1.0    2019-03-16 [1] CRAN (R 4.0.0)
 knitr           * 1.28     2020-02-06 [1] CRAN (R 4.0.0)
 labeling          0.3      2014-08-23 [1] CRAN (R 4.0.0)
 lattice           0.20-41  2020-04-02 [1] CRAN (R 4.0.0)
 lifecycle         0.2.0    2020-03-06 [1] CRAN (R 4.0.0)
 lubridate       * 1.7.9    2020-06-08 [1] CRAN (R 4.0.0)
 magrittr        * 1.5      2014-11-22 [1] CRAN (R 4.0.0)
 MASS              7.3-51.5 2019-12-20 [1] CRAN (R 4.0.0)
 Matrix            1.2-18   2019-11-27 [1] CRAN (R 4.0.0)
 munsell           0.5.0    2018-06-12 [1] CRAN (R 4.0.0)
 mvtnorm           1.1-0    2020-02-24 [1] CRAN (R 4.0.0)
 openssl           1.4.2    2020-06-27 [1] CRAN (R 4.0.0)
 openxlsx        * 4.1.5    2020-05-06 [1] CRAN (R 4.0.0)
 packrat           0.5.0    2018-11-14 [1] CRAN (R 4.0.0)
 pdftools          2.3.1    2020-05-22 [1] CRAN (R 4.0.0)
 pillar            1.4.4    2020-05-05 [1] CRAN (R 4.0.0)
 pkgconfig         2.0.3    2019-09-22 [1] CRAN (R 4.0.0)
 pkgload           1.1.0    2020-05-29 [1] CRAN (R 4.0.0)
 plyr            * 1.8.6    2020-03-03 [1] CRAN (R 4.0.0)
 png               0.1-7    2013-12-03 [1] CRAN (R 4.0.0)
 purrr           * 0.3.4    2020-04-17 [1] CRAN (R 4.0.0)
 qdapRegex         0.7.2    2017-04-09 [1] CRAN (R 4.0.0)
 qpdf              1.1      2019-03-07 [1] CRAN (R 4.0.0)
 R6                2.4.1    2019-11-12 [1] CRAN (R 4.0.0)
 Rcpp              1.0.4.6  2020-04-09 [1] CRAN (R 4.0.0)
 readr             1.3.1    2018-12-21 [1] CRAN (R 4.0.0)
 readxl          * 1.3.1    2019-03-13 [1] CRAN (R 4.0.0)
 rJava             0.9-12   2020-03-24 [1] CRAN (R 4.0.0)
 rlang             0.4.6    2020-05-02 [1] CRAN (R 4.0.0)
 rmarkdown         2.3      2020-06-18 [1] CRAN (R 4.0.0)
 rprojroot         1.3-2    2018-01-03 [1] CRAN (R 4.0.0)
 rstudioapi        0.11     2020-02-07 [1] CRAN (R 4.0.0)
 rvest             0.3.5    2019-11-08 [1] CRAN (R 4.0.0)
 scales            1.1.1    2020-05-11 [1] CRAN (R 4.0.0)
 sessioninfo       1.1.1    2018-11-05 [1] CRAN (R 4.0.0)
 snakecase         0.11.0   2019-05-25 [1] CRAN (R 4.0.0)
 splitstackshape * 1.4.8    2019-04-21 [1] CRAN (R 4.0.0)
 stringi           1.4.6    2020-02-17 [1] CRAN (R 4.0.0)
 stringr         * 1.4.0    2019-02-10 [1] CRAN (R 4.0.0)
 tabulizer       * 0.2.2    2018-06-07 [1] CRAN (R 4.0.0)
 tabulizerjars     1.0.1    2018-05-14 [1] CRAN (R 4.0.0)
 testthat          2.3.2    2020-03-02 [1] CRAN (R 4.0.0)
 textclean       * 0.9.3    2018-07-23 [1] CRAN (R 4.0.0)
 tibble            3.0.1    2020-04-20 [1] CRAN (R 4.0.0)
 tidyr             1.1.0    2020-05-20 [1] CRAN (R 4.0.0)
 tidyselect        1.1.0    2020-05-11 [1] CRAN (R 4.0.0)
 vctrs             0.3.1    2020-06-05 [1] CRAN (R 4.0.0)
 viridisLite       0.3.0    2018-02-01 [1] CRAN (R 4.0.0)
 webshot           0.5.2    2019-11-22 [1] CRAN (R 4.0.0)
 withr             2.2.0    2020-04-20 [1] CRAN (R 4.0.0)
 wrapr           * 2.0.0    2020-03-28 [1] CRAN (R 4.0.0)
 xfun              0.14     2020-05-20 [1] CRAN (R 4.0.0)
 xml2              1.3.2    2020-04-23 [1] CRAN (R 4.0.0)
 zip               2.0.4    2019-09-01 [1] CRAN (R 4.0.0)
```

I agree with others that this does not repeat always. Most likely related to the plot window / graphics device.
Also it is just a warning, and it does not crash my system (yet).

**Comment by r2evans** [comment]
A recent crash/dump gave me a little more info this time,

```
Error in UseMethod("depth") : 
  no applicable method for 'depth' applied to an object of class "NULL"
    x
 1. \-grid:::drawGTree(x)
 2.   +-grid:::postDraw(x)
 3.   \-grid:::postDraw.grob(x)
 4.     +-grid:::popgrobvp(x$vp)
 5.     \-grid:::popgrobvp.viewport(x$vp)
 6.       \-grid::upViewport(depth(vp), recording = FALSE)
 7.         \-grid::depth(path)


==== C stack trace ===============================

	KiUserCallbackDispatcher [0x0x7ff8dce0fbf4+36]
	NtUserDispatchMessage [0x0x7ff8dad11704+20]
	DispatchMessageW [0x0x7ff8dc69ddca+682]
	GA_doevent [0x0x635552f5+229]
	R_setStartTime [0x0x6c72088f+4175]
	Rf_ReplIteration [0x0x6c824fa5+277]
	Rf_ReplIteration [0x0x6c825491+1537]
	run_Rmainloop [0x0x6c825532+82]
	(No symbol) [0x0x40171a]
	(No symbol) [0x0x40158a]
	(No symbol) [0x0x4013c5]
	(No symbol) [0x0x40152b]
	BaseThreadInitThunk [0x0x7ff8dbe66fd4+20]
	RtlUserThreadStart [0x0x7ff8dcdbcec1+33]
```

win10, R-4.0.2, ggplot-3.3.0

**Comment by pmur002** [comment]
Just acknowledging the additional reports:  thanks guys.  Further progress is waiting on myself and/or Simon finding time amongst teaching to take a closer look on a Mac.

**Comment by jlederluis** [comment]
Hi, just wanted to say that we are having the same problem in our repo and I'm following this closely.  Any temporary workarounds are much appreciated!

**Comment by yutannihilation** [comment]
As we haven't seen any new reports for this two years, I think this can be closed now. Thanks again @pmur002 for fixing this!

**Comment by r2evans** [comment]
Thank you @yutannihilation  and @pmur002 , I have tried several times (R-4.1.2, ggplot2-3.3.5) with my previous example (ggplot, then mouse-wheel like a mad-man), and while I still periodically (though less-frequently) see one of

```
### most commonly
Error in UseMethod("depth") : 
  no applicable method for 'depth' applied to an object of class "NULL"
### occasionally
Error in grid.Call.graphics(C_upviewport, as.integer(n)) : 
  cannot pop the top-level viewport ('grid' and 'graphics' output mixed?)
```

I have yet to crash my R.

**Comment by shahnawaz323** [comment]
I have got this error today on debian linux with patchwork graphs. I am using R on VS code and viewing graphs with httpgd in browser. 


```Error in UseMethod("depth") : 
  no applicable method for 'depth' applied to an object of class "NULL"```

**Comment by tfanatik** [comment]
Got this today:
Error in UseMethod("depth") : 
  no applicable method for 'depth' applied to an object of class "NULL"
from this code:
install.packages("ggcorrplot")
library(ggcorrplot)
model.matrix(~0+., data=killings_df) %>% 
  cor(use="pairwise.complete.obs", method = "square") %>% 
  ggcorrplot(show.diag=FALSE, type="lower", lab=TRUE, lab_size=2)

on a Windows10 laptop running R 4.3.1 but working in RStudio 

I went thru this whole thing from the beginning and couldn't see how you guys resolved this but I need help too. Thanks!

**Comment by tfanatik** [comment]
Sorry dudes! I found my error was fixed by adding the method.. the correct method tho! I'm fine, ignore my plea please!

**Comment by strengejacke** [comment]
I found this error recently after updating to ggplot2 3.5.0. I'm using VSCode on windows 11, latest R version. I _guess_ it could be related to the ggplot2 update, because it now happens with code that used to work fine before.

**Comment by teunbrand** [comment]
Can you provide a reprex of what used to work but no longer does?

**Comment by strengejacke** [comment]
I'll try, and if I have one, I would open a new issue.

**Comment by sofiabiffi** [comment]
> I found this error recently after updating to ggplot2 3.5.0. I'm using VSCode on windows 11, latest R version. I _guess_ it could be related to the ggplot2 update, because it now happens with code that used to work fine before.

I am also in this exact situation since the latest update, the first couple plots I run in a new session work and then I get the depth issue error and the plots appear completely misdrawn

**Comment by teunbrand** [comment]
Paul Murrell previously traced a very similar error to issues with the graphics device. I'm not sure what graphics device VSCode uses, but seems likely that the error orginates there.

**Comment by sofiabiffi** [comment]
> Paul Murrell previously traced a very similar error to issues with the graphics device. I'm not sure what graphics device VSCode uses, but seems likely that the error originates there.

Ah I see, I'll dig deeper then. Thank you so much
